pkg/quarterdeck/apikeys.go

```diff
@@ -434,41 +445,21 @@ func (s *Server) APIKeyPermissions(c *gin.Context) {
 	}
 
 	// Fetch all eligible API key claims
-	out := make([]string, 0, 7)
-	if tx, err = db.BeginTx(c.Request.Context(), &sql.TxOptions{ReadOnly: true}); err != nil {
+	if allPermissions, err = models.GetAPIKeyPermissions(c.Request.Context()); err != nil {
 		c.Error(err)
 		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not fetch apikey permissions"))
 		return
 	}
-	defer tx.Rollback()
-
-	if rows, err = tx.Query("SELECT name FROM permissions WHERE allow_api_keys=true"); err != nil {
-		c.Error(err)
-		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not fetch apikey permissions"))
-		return
-	}
-	defer rows.Close()
-
-	for rows.Next() {
-		var permission string
-		if err = rows.Scan(&permission); err != nil {
-			c.Error(err)
-			c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not fetch apikey permissions"))
-			return
-		}
-		out = append(out, permission)
-	}
 
 	// Filter other permissions based on the user's claims.
-	outf := make([]string, 0, len(out))
-	for _, permission := range out {
-		// TODO: we'll need a better way to identify permissions that both the user and the API key can have.
-		if perms.InGroup(permission, perms.PrefixTopics) || perms.InGroup(permission, perms.PrefixMetrics) {
-			if !claims.HasPermission(permission) {
-				// Do not return this permission
-				continue
-			}
+	outf := make([]string, 0, len(allPermissions))
+	for _, permission := range allPermissions {
+		if perms.UserKeyPermission(permission) && !claims.HasPermission(permission) {
+			// Do not return this permission
+			continue
 		}
+
+		// Build sorted return list for the user (note that the db query returns a sorted array)
 		outf = append(outf, permission)
```

Comment:
Since the check wasn't necessary I returned this to a `[]string` and added an `ORDER BY` clause in the query to make sure they were sorted.